### PR TITLE
Add multi-backend testing for Keystone SAML Mellon

### PR DESCRIPTION
**Pull request by ionutbalutoiu**
The new updated tests rely on new testing bundles deployed with
two local IdPs via the Juju charm https://jaas.ai/u/ionutbalutoiu/test-saml-idp.

### CODE REVIEW

File: zaza/openstack/charm_tests/saml_mellon/tests.py [hunk 1]
```diff
@@ -20,28 +20,28 @@
...
 
 
 class FailedToReachIDP(Exception):
     """Custom Exception for failing to reach the IDP."""
 
-    pass
 
-
-class CharmKeystoneSAMLMellonTest(BaseKeystoneTest):
+class BaseCharmKeystoneSAMLMellonTest(BaseKeystoneTest):
```
Comment (ajkavanagh) [inline zaza/openstack/charm_tests/saml_mellon/tests.py:30]: Changing the name from `CharmKeystoneSAMLMellonTest` to `CharmKeystoneSAMLMellonIDP1Test` will, I think, break stable versions of the charm tests.  We tend to have to leave the old names alone and add new ones only so as to not break existing stable charms.

**Review by ajkavanagh — CHANGES_REQUESTED** [review]
Looks good apart from the test name change for stable charms / existing tests (e.g. from a tests.yaml).  As the Test names are essentially the API to the testing library they tend to be "forever".

File: zaza/openstack/charm_tests/saml_mellon/tests.py (hunk 1 above)
Comment (ionutbalutoiu) [reply zaza/openstack/charm_tests/saml_mellon/tests.py:30]: @ajkavanagh

Makes sense. I'll update the PR, and make sure to keep the backwards compatibility.

**Comment by ionutbalutoiu** [comment]
> Looks good apart from the test name change for stable charms / existing tests (e.g. from a tests.yaml). As the Test names are essentially the API to the testing library they tend to be "forever".

I updated the PR and took care of the backwards compatibility. I also validated these changes with a `func-smoke` against:
* Latest master with my bundles changes from https://review.opendev.org/c/openstack/charm-keystone-saml-mellon/+/770056
* The `stable/20.10` charm release.

Everything went fine on my local Zaza environment.

There's some duplicate code in the `zaza/openstack/charm_tests/saml_mellon/tests.py` between the old `CharmKeystoneSAMLMellonTest` class (which uses upstream `samltest.id`), and the new `BaseCharmKeystoneSAMLMellonTest` class (which uses the local `test-saml-idp`).

I left it like this on purpose, since we can remove the old testing class, once we fully drop the `samltest.id` dependency (even on stable charm releases).

**Review by AurelienLourot — CHANGES_REQUESTED** [review]
I haven't reviewed yet, just posting this comment first, but I'll circle back to it soon, thanks!

File: zaza/openstack/charm_tests/saml_mellon/tests.py [hunk 2]
```diff
@@ -156,3 +156,177 @@ def _do_redirect_check(url, region, idp_expect, horizon_expect):
         # We may need to try/except to allow horizon to build its pages
         _do_redirect_check(url, region, idp_expect, horizon_expect)
         logging.info("SUCCESS")
+
```
[the remark this answers is not in the document]
Comment (ionutbalutoiu) [reply zaza/openstack/charm_tests/saml_mellon/tests.py:159]: @AurelienLourot Sounds good. I'll wait for your review, and address all the comments at once.

[the remark this answers is not in the document]
Comment (ionutbalutoiu) [reply zaza/openstack/charm_tests/saml_mellon/tests.py:161]: Added a comment about the deprecation of the old testing class using `samltest.id` dependency.

File: zaza/openstack/charm_tests/saml_mellon/tests.py [hunk 3]
```diff
@@ -156,3 +158,177 @@ def _do_redirect_check(url, region, idp_expect, horizon_expect):
...
+    @classmethod
+    def setUpClass(cls,
+                   application_name="keystone-saml-mellon",
+                   test_saml_idp_app_name="test-saml-idp",
+                   horizon_idp_option_name="myidp_mapped",
+                   horizon_idp_display_name="myidp via mapped"):
+        """Run class setup for running Keystone SAML Mellon charm tests."""
+        super(BaseCharmKeystoneSAMLMellonTest, cls).setUpClass()
+        cls.application_name = application_name
```
Comment (AurelienLourot) [inline zaza/openstack/charm_tests/saml_mellon/tests.py:174]: This note was useful, can we keep it?

```
        # Note: The BaseKeystoneTest class sets the application_name to
        # "keystone" which breaks keystone-saml-mellon actions. Explicitly set
        # application name here.
```

**Review by AurelienLourot — CHANGES_REQUESTED** [review]
LGTM in general, thanks!

File: zaza/openstack/charm_tests/saml_mellon/setup.py [hunk 4]
```diff
@@ -161,3 +165,88 @@ def attach_saml_resources(application="keystone-saml-mellon"):
...
+        test_saml_idp_app_name).pop()
+    action_result = zaza.model.run_action(
+        test_saml_idp_unit.name, 'get-idp-metadata')
+    idp_metadata = action_result.data['results']['output']
+
+    with tempfile.NamedTemporaryFile(mode='w', suffix='.xml') as fp:
+        fp.write(idp_metadata)
+        fp.flush()
+        zaza.model.attach_resource(
+            keystone_saml_mellon_app_name, "idp-metadata", fp.name)
```
Comment (ajkavanagh) [inline zaza/openstack/charm_tests/saml_mellon/setup.py:183]: There's a pattern developing here (this is repeated 3 times).  Would breaking out a helper that does this and attaches the resource (or even add a new helper), help with the clarity and intention of the code.  e.g.

```python
zaza.utils.attach_resource(app, "idp-metadata', idp_metadata, ".xml")
```

File: zaza/openstack/charm_tests/saml_mellon/tests.py [hunk 5]
```diff
@@ -156,3 +158,177 @@ def _do_redirect_check(url, region, idp_expect, horizon_expect):
...
+        cls.current_release = openstack_utils.get_os_release()
+        cls.FOCAL_USSURI = openstack_utils.get_os_release("focal_ussuri")
+
+    def test_run_get_sp_metadata_action(self):
+        """Validate the get-sp-metadata action."""
+        unit = zaza.model.get_units(self.application_name)[0]
+        if self.vip:
+            ip = self.vip
+        else:
+            ip = unit.public_address
```
Comment (ajkavanagh) [inline zaza/openstack/charm_tests/saml_mellon/tests.py:188]: I tend to prefer (but don't change it if you prefer the current format):

```python
ip = self.vip if self.vip else unit.public_address
```

File: zaza/openstack/charm_tests/saml_mellon/tests.py [hunk 6]
```diff
@@ -156,3 +158,177 @@ def _do_redirect_check(url, region, idp_expect, horizon_expect):
...
+
+        action = zaza.model.run_action(unit.entity_id, self.action)
+        if "failed" in action.data["status"]:
+            raise Exception(
+                "The action failed: {}".format(action.data["message"]))
```
Comment (ajkavanagh) [inline zaza/openstack/charm_tests/saml_mellon/tests.py:193]: This is test code.  Why not:

```python
self.assertNotIn("failed", action.data["status"], msg="The action failed ....")
```

File: zaza/openstack/charm_tests/saml_mellon/tests.py [hunk 7]
```diff
@@ -156,3 +158,177 @@ def _do_redirect_check(url, region, idp_expect, horizon_expect):
...
+
+        output = action.data["results"]["output"]
+        root = etree.fromstring(output)
+        for item in root.items():
+            if "entityID" in item[0]:
+                assert ip in item[1]
```
Comment (ajkavanagh) [inline zaza/openstack/charm_tests/saml_mellon/tests.py:199]:
```python
self.assertIn(ip, item[1])
```

File: zaza/openstack/charm_tests/saml_mellon/tests.py [hunk 8]
```diff
@@ -156,3 +158,177 @@ def _do_redirect_check(url, region, idp_expect, horizon_expect):
...
+
+        for appt in root.getchildren():
+            for elem in appt.getchildren():
+                for item in elem.items():
+                    if "Location" in item[0]:
+                        assert ip in item[1]
```
Comment (ajkavanagh) [inline zaza/openstack/charm_tests/saml_mellon/tests.py:205]: As above.

File: zaza/openstack/charm_tests/saml_mellon/tests.py [hunk 9]
```diff
@@ -156,3 +158,177 @@ def _do_redirect_check(url, region, idp_expect, horizon_expect):
...
+        if horizon_vip:
+            horizon_ip = horizon_vip
+        else:
+            unit = zaza.model.get_units("openstack-dashboard")[0]
+            horizon_ip = unit.public_address
+
+        if self.tls_rid:
+            proto = "https"
+        else:
+            proto = "http"
```
Comment (ajkavanagh) [inline zaza/openstack/charm_tests/saml_mellon/tests.py:229]: One line if-else possible if preferred.

File: zaza/openstack/charm_tests/saml_mellon/tests.py [hunk 10]
```diff
@@ -156,3 +158,177 @@ def _do_redirect_check(url, region, idp_expect, horizon_expect):
...
+                msg = "FAILURE code={} text={}".format(response, response.text)
+                # Raise a custom exception.
+                raise FailedToReachIDP(msg)
+
+            idp_url = ("http://{0}/simplesaml/"
+                       "module.php/core/loginuserpass.php").format(idp_address)
+
+            # Validate that we were redirected to the proper IdP
+            assert response.url.startswith(idp_url)
+            assert idp_url in response.text
```
Comment (ajkavanagh) [inline zaza/openstack/charm_tests/saml_mellon/tests.py:296]: This is a big function to be embedded in another.  Please break it out and add `@staticmethod` to it, plus docstring with purpose, parameters and types.

**Review by ajkavanagh — CHANGES_REQUESTED** [review]
Please see the single comments; apologies for not starting a review.   I'd already proceeding with the single comments and it was too late by then!

File: zaza/openstack/charm_tests/saml_mellon/setup.py [hunk 11]
```diff
@@ -81,56 +81,60 @@
...
     try:
-        keystone_client.federation.protocols.get(IDP, PROTOCOL_NAME)
+        keystone_client.federation.protocols.get(idp_name, PROTOCOL_NAME)
     except keystoneauth1.exceptions.http.NotFound:
         keystone_client.federation.protocols.create(
-            PROTOCOL_NAME, mapping=MAP_NAME, identity_provider=idp)
+            PROTOCOL_NAME, mapping=map_name, identity_provider=idp)
 
 
 def attach_saml_resources(application="keystone-saml-mellon"):
```
[the remark this answers is not in the document]
Comment (ionutbalutoiu) [reply zaza/openstack/charm_tests/saml_mellon/setup.py:140]: Yes. I'll add the commend.

File: zaza/openstack/charm_tests/saml_mellon/setup.py [hunk 12]
```diff
@@ -161,3 +165,88 @@ def attach_saml_resources(application="keystone-saml-mellon"):
         fp.flush()
         zaza.model.attach_resource(
             application, sp_signing_keyinfo_name, fp.name)
+
+
+def _attach_saml_resources_local_idp(keystone_saml_mellon_app_name=None,
+                                     test_saml_idp_app_name=None):
+    """Attach resources to the Keystone SAML Mellon and the local IdP."""
+    test_saml_idp_unit = zaza.model.get_units(
+        test_saml_idp_app_name).pop()
```
[the remark this answers is not in the document]
Comment (ionutbalutoiu) [reply zaza/openstack/charm_tests/saml_mellon/setup.py:174]: In fact, I only need to run the `get-idp-metadata` action, and I noticed that we have `run_action_on_leader` and I don't even need to fetch an unit. I'll use that.

File: zaza/openstack/charm_tests/saml_mellon/setup.py (hunk 4 above)
Comment (ionutbalutoiu) [reply zaza/openstack/charm_tests/saml_mellon/setup.py:183]: Good suggestion. Will implement it.

File: zaza/openstack/charm_tests/saml_mellon/setup.py [hunk 13]
```diff
@@ -161,3 +165,88 @@ def attach_saml_resources(application="keystone-saml-mellon"):
...
+            keystone_saml_mellon_app_name, 'sp-private-key', fp.name)
+
+    with tempfile.NamedTemporaryFile(mode='w', suffix='.xml') as fp:
+        fp.write(SP_SIGNING_KEY_INFO_XML_TEMPLATE.format(cert))
+        fp.flush()
+        zaza.model.attach_resource(
+            keystone_saml_mellon_app_name, 'sp-signing-keyinfo', fp.name)
+
+    keystone_saml_mellon_unit = zaza.model.get_units(
+        keystone_saml_mellon_app_name).pop()
```
[the remark this answers is not in the document]
Comment (ionutbalutoiu) [reply zaza/openstack/charm_tests/saml_mellon/setup.py:203]: Same solution as above. I can use `runa_action_on_leader`.

File: zaza/openstack/charm_tests/saml_mellon/tests.py (hunk 3 above)
Comment (ionutbalutoiu) [reply zaza/openstack/charm_tests/saml_mellon/tests.py:174]: Will add it back.

File: zaza/openstack/charm_tests/saml_mellon/tests.py (hunk 5 above)
Comment (ionutbalutoiu) [reply zaza/openstack/charm_tests/saml_mellon/tests.py:188]: I like this as well. I'll change it.

File: zaza/openstack/charm_tests/saml_mellon/tests.py (hunk 6 above)
Comment (ionutbalutoiu) [reply zaza/openstack/charm_tests/saml_mellon/tests.py:193]: Will change it.

File: zaza/openstack/charm_tests/saml_mellon/tests.py (hunk 10 above)
Comment (ionutbalutoiu) [reply zaza/openstack/charm_tests/saml_mellon/tests.py:296]: This function was carried on from the previous implementation.

I agree with you. I'll clean it up.

**Comment by ionutbalutoiu** [comment]
@ajkavanagh @AurelienLourot

Thank-you for the reviews. I addressed all the commends. Please check the updated PR when you have a chance.

**Review by ajkavanagh — APPROVED** [review]
LGTM - thanks for making the  changes, particularly the code-gardening around the embedded function; much clearer now.

**Review by AurelienLourot — APPROVED** [review]
Perfect, thanks! Happy to land this when the corresponding gerrit review turns green. ( https://review.opendev.org/c/openstack/charm-keystone-saml-mellon/+/770056 )

**Comment by AurelienLourot** [comment]
Note: the `.decode()` call has been moved, this is what has changed in the last push-force. Still LGTM. @ionutbalutoiu can you please not push-force? Because we can't easily spot here in the browser what changed and this requires special steps to figure it out. Thanks!

**Comment by ionutbalutoiu** [comment]
> Note: the `.decode()` call has been moved, this is what has changed in the last push-force. Still LGTM. @ionutbalutoiu can you please not push-force? Because we can't easily spot here in the browser what changed and this requires special steps to figure it out. Thanks!

Oh, sorry about `force-push`. I tend to have a single clean commit. But I realize that it can get confusing when the changes are reviewed already.

The `decode()` had to be removed from there, because the helper function is meant to get a `str` (not an encoded `str`).

After I got your approvals, I quickly ran a local Zaza run and noticed that the function was failing when it got `str`, because it tried to `decode()` it. The `decode()` needs to be called before using the helper (if needed).

That's the full story of the most recent change. Our local Zaza run went through without any problems after fixing this.

**Review by AurelienLourot — APPROVED** [review]

**Pull request merged**